design-documents/testing/coding-standards/phpcompatibility.md

```diff
@@ -0,0 +1,28 @@
+# PHPCompatibility in Code Style Ruleset
+
+## Goal
+
+Utilize PHPCompatibility tool in `phpcs` rules, so that rules specific to PHP version are validated.
+
+Requirements include determining PHP version from `composer.json` of the Magento project or Magento module, so validation against correct version happens automatically.
+This requirement though introduces additional complexity to the ruleset/sniff implementation:
+- `composer.json` may be absent
+- it requires custom bootstrapping logic for `phpcs`
+- it's not clear whether we should analyze different files differently in case they are from different modules and those modules hav different PHP constraints
+
+Based on the above it is proposed to put the responsibility of specifying PHP version to the user (owner of the project).
+
+## Implementation Details
+
+In https://github.com/magento/magento-coding-standard repository:
+1. Add `phpcompatibility/php-compatibility` to Composer dependencies
+2. Add necessary configuration to the ruleset, so that PHPCompatibility ruleset is included
+   2.1. Add `<config name="installed_paths" value="./../../../../../vendor/phpcompatibility/php-compatibility" />`
+   2.2. Add `<rule ref="PHPCompatibility"/>`
+
+There are two options for https://github.com/magento/magento2 repository:
+1. Add `<config name="testVersion" value="7.1.3-7.3.0"/>` to Magento static tests ruleset
+   1.1. Different branches will have different PHP versions
+   1.2. Drawback: PHP versions in the ruleset should be updated every time Magento changes supported PHP versions
+2. Determine PHP version in `\Magento\Test\Php\LiveCodeTest::testCodeStyle()` based on the project's `composer.json` and pass it as an argument to `phpcs`
```

Comment:
I'd argue that using the namespace `\Magento\Test` is not such a good idea as it may be confused with the namespace `\Magento\Tests` in which the automatic tests are defined.
The same argument stands valid for the class name `LiveCodeTest` and the method name `testCodeStyle` as they are too close to the naming convention for test classes and cases.  

If I understand you right, you'd propose to use this class as a kickstarter  for `phpcs`. This would indeed have the drawback that IDEs would not take into account any PHP version and it also has the drawback, that poeple need to call another command than `phpcs` to have PHP versions taken into account.  
In order to eliminate those drawbacks, I'd suggest to use a config [bootstrap](https://github.com/squizlabs/PHP_CodeSniffer/wiki/Advanced-Usage#using-a-bootstrap-file) to execute class in the ruleset. This can be specified as an argument to `phpcs`and can thus be contained in the `ruleset.xml`:
```xml
  <arg name="bootstrap" value="php-compatibility-bootstrap.php" />
```

In turn I'd suggest to rename `\Magento\Test\Php\LiveCodeTest::testCodeStyle()` to `\Magento\Bootstrap\PhpCompatibility::execute()` and have the bootstrap manipulate the configuration settings.

The file `php-compatibility-bootstrap.php` would then contain the following code:
```php
<?php
/**
 * Manipulate config settings for package 'phpcompatibility/php-compatibility'
 *
 * @var \PHP_CodeSniffer\Runner $this
 */
use Magento\Bootstrap\PHPCompatibility;

$bootstrap = new PHPCompatibility($this);
$bootstrap->execute();
```

The variable `$this` should be available in the context of the bootstrap file as the file is being `include`ed in `\PHP_CodeSniffer\Runner::run()`.

Reply:
The class I mentioned already exists and it's used for starting `phpcs` in our builds.
My idea is:
- automate PHP version detection for our builds
- keep implementation as simple as possible for 3rd-party. The idea here is that we just include in our docs something like "Now we have PHPCompatibility rules included, please go read about it in its official documentation". Then people can follow PHPCompatibility docs and either specify or not specify PHP version. I'd really like to avoid overcomplicating the implementation and especially the case when it behaves non-standard or does not allow to configure it in standard way (e.g., due to a bug or incorrect assumption, it becomes impossible to reconfigure PHP version and it will be always auto-calculated).

Reply:
Ok, I wasn't aware that the class already exists nor could I find it in the repo `magento-coding-standard`. Obviousel I searched in the wrong place. Sorry for the inconvenience.